File: app/src/main/java/me/blog/korn123/easyphotomap/adapters/SearchItemAdapter.java
package me.blog.korn123.easyphotomap.adapters;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import org.apache.commons.io.FilenameUtils;

import java.util.ArrayList;

import me.blog.korn123.easyphotomap.R;
import me.blog.korn123.easyphotomap.constants.Constant;
import me.blog.korn123.easyphotomap.models.PhotoMapItem;
import me.blog.korn123.easyphotomap.utils.BitmapUtils;
import me.blog.korn123.easyphotomap.utils.CommonUtils;

/**
 * Created by CHO HANJOONG on 2016-07-20.
 */
public class SearchItemAdapter extends ArrayAdapter<PhotoMapItem> {

    private final Context context;
    private final ArrayList<PhotoMapItem> entities;
    private final int layoutResourceId;
    private Activity activity;

    public SearchItemAdapter(Context context, Activity activity, int layoutResourceId, ArrayList<PhotoMapItem> entities) {
        super(context, layoutResourceId, entities);
        this.context = context;
        this.entities = entities;
        this.layoutResourceId = layoutResourceId;
        this.activity = activity;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View row = convertView;
        ViewHolder holder = null;

        if(row == null) {
            LayoutInflater inflater = ((Activity)context).getLayoutInflater();
            row = inflater.inflate(layoutResourceId, parent, false);
            holder = new ViewHolder();
            holder.textView1 = (TextView)row.findViewById(R.id.text1);
            holder.textView2 = (TextView)row.findViewById(R.id.text2);
            holder.textView3 = (TextView)row.findViewById(R.id.text3);
            holder.imageView1 = (ImageView)row.findViewById(R.id.image1);
            row.setTag(holder);
        } else {
            holder = (ViewHolder)row.getTag();
        }

        PhotoMapItem imageEntity = entities.get(position);
        Bitmap bitmap = null;
        String fileName = FilenameUtils.getName(imageEntity.imagePath);
        bitmap = BitmapUtils.decodeFile(activity, Constant.WORKING_DIRECTORY + fileName + ".thumb");
        holder.textView1.setText(imageEntity.info);
        holder.textView2.setText(imageEntity.date.toString());
        holder.textView3.setText(imageEntity.imagePath);
        holder.imageView1.setImageBitmap(bitmap);
        return row;
    }

    static class ViewHolder {
        TextView textView1;
        TextView textView2;
        TextView textView3;
        ImageView imageView1;
    }
}

File: app/src/main/java/me/blog/korn123/easyphotomap/adapters/SearchItemAdapter.kt
package me.blog.korn123.easyphotomap.adapters

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.TransitionDrawable
import android.os.AsyncTask
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ImageView
import android.widget.TextView
import me.blog.korn123.easyphotomap.R
import me.blog.korn123.easyphotomap.constants.Constant
import me.blog.korn123.easyphotomap.models.PhotoMapItem
import me.blog.korn123.easyphotomap.utils.BitmapUtils
import me.blog.korn123.easyphotomap.utils.CommonUtils
import org.apache.commons.io.FilenameUtils
import java.util.*

/**
 * Created by CHO HANJOONG on 2016-07-20.
 */
class SearchItemAdapter(private val activity: Activity,
                        private val listPhotoMapItem: ArrayList<PhotoMapItem>,
                        private val onItemClickListener: AdapterView.OnItemClickListener,
                        private val onItemLongClickListener: AdapterView.OnItemLongClickListener
) : RecyclerView.Adapter<SearchItemAdapter.ViewHolder>() {
    private val layoutInflater = activity.layoutInflater

    private fun createViewHolder(layoutType: Int, parent: ViewGroup?): SearchItemAdapter.ViewHolder {
        val view = layoutInflater.inflate(layoutType, parent, false)
        return SearchItemAdapter.ViewHolder(view as ViewGroup)
    }

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        val photoMapItem: PhotoMapItem = listPhotoMapItem[position]
        holder?.run {
            text1?.text = photoMapItem.info
            text2?.text = photoMapItem.date
            text3?.text = photoMapItem.imagePath
            val fileName = FilenameUtils.getName(photoMapItem.imagePath)
            SearchItemAdapter.ThumbnailTask(activity, position, holder).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, Constant.WORKING_DIRECTORY + fileName + ".thumb")
        }

        holder?.let {
            it.itemView.setOnClickListener { item ->
                onItemClickListener.onItemClick(null, item, it.adapterPosition, it.itemId)
            }
            it.itemView.setOnLongClickListener { item ->
                onItemLongClickListener.onItemLongClick(null, item, it.adapterPosition, it.itemId)
            }
        }
    }

    override fun getItemCount(): Int = listPhotoMapItem.size

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder = createViewHolder(R.layout.item_search, parent)

    fun getItem(position: Int): PhotoMapItem = listPhotoMapItem[position]

    private class ThumbnailTask(val activity: Activity, val position: Int, val holder: SearchItemAdapter.ViewHolder) : AsyncTask<String, Void, Bitmap>() {
        override fun doInBackground(vararg params: String): Bitmap? {
            val filePath = params[0]
//            val widthHeight = CommonUtils.dpToPixel(activity, 45f)
            var resized: Bitmap? = null
            if (holder.position == position) {
                var bitmap = BitmapUtils.decodeFile(activity, filePath)
//                resized = Bitmap.createScaledBitmap(bitmap, widthHeight, widthHeight, true)
                resized = bitmap
            } else {
                this.cancel(true)
            }
            return resized
        }

        override fun onPostExecute(bitmap: Bitmap) {
            if (holder.position == position) {
                val td = TransitionDrawable(arrayOf(ColorDrawable(Color.TRANSPARENT), BitmapDrawable(activity.resources, bitmap)))
                holder.image1?.setImageDrawable(td)
                td.startTransition(1000)
            }
        }
    }

    class ViewHolder(val parent: ViewGroup?) : RecyclerView.ViewHolder(parent) {
        var text1: TextView? = null
        var text2: TextView? = null
        var text3: TextView? = null
        var image1: ImageView? = null
        init {
            parent?.let {
                text1 = parent.findViewById(R.id.text1)
                text2 = parent.findViewById(R.id.text2)
                text3 = parent.findViewById(R.id.text3)
                image1 = parent.findViewById(R.id.image1)
            }
        }
    }
}
